kyo-prelude/shared/src/main/scala/kyo2/Layer.scala

```diff
@@ -0,0 +1,105 @@
+package kyo2
+
+import Layer.internal.*
+import java.util.concurrent.ConcurrentHashMap as JConcurrentHashMap
+import kyo.Tag
+import kyo2.kernel.Reducible
+import scala.annotation.targetName
+
+abstract class Layer[+Out, -S]:
+    self =>
+
+    final infix def to[Out2, S2, In2](that: Layer[Out2, Env[In2] & S2]): Layer[Out2, S & S2]          = To(self, that)
+    final infix def and[Out2, S2](that: Layer[Out2, S2]): Layer[Out & Out2, S & S2]                   = And(self, that)
+    final infix def using[Out2, S2, In2](that: Layer[Out2, Env[In2] & S2]): Layer[Out & Out2, S & S2] = self and (self to that)
+
+end Layer
+
+object Layer:
+
+    extension [In, Out, S](layer: Layer[Out, Env[In] & S])
+        def run[R](using reduce: Reducible[Env[In]]): TypeMap[Out] < (S & reduce.SReduced & Memo) =
+            reduce(doRun(layer))
+
+    val empty: Layer[Any, Any] = FromKyo { () => TypeMap.empty }
+
+    def apply[A: Tag, S](kyo: => A < S): Layer[A, S] =
+        FromKyo { () =>
+            kyo.map { result => TypeMap(result) }
+        }
+
+    def from[A: Tag, B: Tag, S](f: A => B < S): Layer[B, Env[A] & S] =
+        apply {
+            Env.get[A].map(f)
+        }
+
+    def from[A: Tag, B: Tag, C: Tag, S](f: (A, B) => C < S): Layer[C, Env[A & B] & S] =
+        apply {
+            Kyo.zip(Env.get[A], Env.get[B]).map { case (a, b) => f(a, b) }
+        }
+
+    def from[A: Tag, B: Tag, C: Tag, D: Tag, S](f: (A, B, C) => D < S): Layer[D, Env[A & B & C] & S] =
+        apply {
+            Kyo.zip(Env.get[A], Env.get[B], Env.get[C])
+                .map { case (a, b, c) => f(a, b, c) }
+        }
+
+    def from[A: Tag, B: Tag, C: Tag, D: Tag, E: Tag, S](f: (A, B, C, D) => E < S): Layer[E, Env[A & B & C & D] & S] =
+        apply {
+            Kyo.zip(Env.get[A], Env.get[B], Env.get[C], Env.get[D]).map { case (a, b, c, d) => f(a, b, c, d) }
+        }
+
+    @targetName("fromPure1")
```

Comment:
I had to introduce these methods because the compiler can't infer the result type as `<` from a pure function anymore since the lifting of pure values into `<` now happens via an implicit conversion.

Reply:
Hmmm. I would like to avoid this if possible. I wonder if a private implicit in this scope might change things...

Reply:
I can't see a way to avoid it, the trouble is on the callers of `Layer.from`. The compiler isn't able to infer the implicit conversion to lift the result type of the provided functions. If you removed these methods, `LayerTest` shows the issue.